### DISCUSSION TITLE
Using Template Variables in `if` condition

### DISCUSSION THREAD

**Discussion by themitigater** [apache/airflow#34520] [General]
Hi,
I'm using Airflow 2.5.3 (Staging Env) with GCP Cloud Composer & Dataflow to run Apache BEAM Jobs

We're in the process of migrating to Airflow 2.0 and one of the things that I'm stuck on is replacing `dag_run.conf` with params since `dag_run.conf` has been deprecated and thus breaks our manual trigger functionality, I'm using params to achieve this.

**Previous Code:**
```
my_options = {
    "clickhouseURL": f"jdbc:clickhouse://{conn.host}:8123/idfy_dlk_master?user={conn.login}&password={conn.password}",
    "startTime": '{{ dag_run.conf["startTime"] or (data_interval_start - macros.timedelta(minutes=5)) }}',
    "minutes": '{{ dag_run.conf["minutes"] or 0 }}',
    "hours": '{{ dag_run.conf["hours"] or 1 }}',
    "days": '{{ dag_run.conf["days"] or 0 }}',
}
```

**New Code:**
```
def analytical_pipeline_params(env_prefix, base_hook, start_time, minutes, hours, days):
    """Builds input options for Analytical Pipeline"""
    conn = base_hook.get_connection(f"{env_prefix}clickhouse_default")

    return {
      "clickhouseURL": Param(f"jdbc:clickhouse://{conn.host}:8123/idfy_dlk_master?user={conn.login}&password={conn.password}", 
                            type="string",
                            title="Clickhouse URL"),
      "startTime": Param(start_time,
                        #  Change to date-time with Version 2.7.0
                        type="string",
                        title="Start Time",
                        description="Start Time variable to be passed to the BEAM Pipeline, format: 2023-09-08T06:05:00+00:00"),
      "minutes": Param(minutes, type="integer", title="Minutes"),
      "hours": Param(hours, type="integer", title="Hours"),
      "days": Param(days, type="integer", title="Days"),
  }

------ DAG File ------
params = analytical_pipeline_params(ENV_PREFIX,
        BaseHook,
        DEFAULT_START_TIME,
        0,1,0
        )

START_TIME = "{{ data_interval_start - macros.timedelta(minutes=5)}}"

my_options = {
        "clickhouseURL": '{{params.clickhouseURL}}',
        "minutes": '{{ params.minutes }}',
        "hours": '{{ params.hours }}',
        "days": '{{ params.days }}',
        "type_params": type('{{params.startTime}}'),
        "type_default": type(DEFAULT_START_TIME),
        "DEFAULT_START_TIME": DEFAULT_START_TIME,
        "matching": "Yes" if params_start_time == DEFAULT_START_TIME else "No"
    }

if '{{params.startTime}}' == DEFAULT_START_TIME:
        my_options["startTime"] = START_TIME
    else:
        my_options["startTime"] = "{{params.startTime}}"

```

Now I of the things I want to achieve if set `my_options['startTime']` using `data_interval_start` if it's an automated run else pick startTime from params..

Above code if me trying to check if params.startTime is equal to my default, but `if '{{params.startTime}}' == DEFAULT_START_TIME` always returns False. Can someone help me understand if template variables can be used in `if...else` blocks? If not, then is there any flag available which says that this is a Manual Run?

**Comment by Taragolis** [comment]
> if '{{params.startTime}}' == DEFAULT_START_TIME always returns False. Can someone help me understand if template variables can be used in if...else blocks? 

No you can't use un-rendered templates in if condition, because until it rendered it is a just a string and render happen only before task started, so in this case you literally compare `DEFAULT_START_TIME` with string `{{params.startTime}}`.
You need to write your conditions inside of the Jinja template, see for reference: https://jinja.palletsprojects.com/en/3.0.x/templates/#if

> If not, then is there any flag available which says that this is a Manual Run?

You could get `dag_run` from the operator's context, DagRun object also have a parameter `run_type` which right now (Airflow 2.7.1) could be one of this values 
https://github.com/apache/airflow/blob/827962878e6fb39e014639d83cff7d0881595ecb/airflow/utils/types.py#L47-L53

So if it manual run, then it should be "manual".

You also could use this Jinja template (I do not test it):

```python
"{{% if dag_run.run_type != 'manual' }}{{ data_interval_start }}{{% else %}}{{ params.startTime }}{{% endif %}}"
```

**Reply by themitigater** [reply]
Thanks. I will test this out.

**Reply by hrishikeshbihani** [reply]
```
"startTime": "{{ data_interval_start - macros.timedelta(minutes=5)  if dag_run.run_type != 'manual' else params.startTime }}",
```
Above config worked for me
